icefoganalytics/travel-authorization
# Rebuild Front-End Authorization Pipeline

Fixes https://github.com/icefoganalytics/travel-authorization/issues/195

Relates to:
- https://github.com/icefoganalytics/travel-authorization/issues/192
- https://github.com/icefoganalytics/internal-data-portal

# Context

**Is your feature request related to a problem? Please describe.**
The current front-end authorization pipeline is very hard to debug.
The auth0 integration is several versions out of date, and the code is spread across multiple files.

**Describe the solution you'd like**
I want to upgrade the Auth0 integration to the latest version.
I want to rebuild the front-end authorization pipeline so it looks like the one in use in the [Internal Data Portal](https://github.com/icefoganalytics/internal-data-portal) project.

**Describe alternatives you've considered**
Keep trying to debug the current system.

**Additional context**
Files that use authorization that I've found so far:
- `web/src/main.js`
- `web/src/api/http-client.js`
- `web/src/auth/axiosAPIConfig.js`
- `web/src/layouts/Layout.vue`
- `web/src/pages/AuthLoadingOverlay.vue`
- `web/src/router/index.js`
- `web/src/auth/index.js`
- `web/src/store/index.js`
- `web/src/store/jwt.js`
- `web/src/modules/authentication/views/SignIn.vue`
- `web/src/auth/authGuard.js`

Patterns that will need replacing:
- `$auth`
- `secureXYZ` - all request should be secure by default, this can be replaced by the http-client code.

# Implementation

Now authorization occurs in these places only:
1. `web/src/App.vue` - via `useAuth0` from `web/src/plugins/auth0-plugin.js` and `useCurrentUser` from `web/src/use/use-current-user.js`
2. `web/src/plugins/auth0-plugin.js` - on app load
3. `web/src/router.js` - via `web/src/utils/auth-guard.js`
4. `web/src/api/http-client.js` via `web/src/plugins/auth0-plugin.js`

Rework auth0 plugin with code from https://github.com/auth0/auth0-vue/blob/efe1d39a0244fd072a527df97f87ab355501851d/src/plugin.ts
Add authGuard from https://github.com/auth0/auth0-vue/blob/efe1d39a0244fd072a527df97f87ab355501851d/src/guard.ts#L16

Front-end routing is now "Secure by Default"(tm). To disable security on a route set `meta: { requiresAuth: false }`

# Screenshots

[Screencast from 2024-06-13 16-44-56.webm](https://github.com/icefoganalytics/travel-authorization/assets/23045206/68f4a3a9-92c0-45f9-901e-b8fad9d718a4)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
5. Boot the app via `dev up`
6. Log in to the app at http://localhost:8080
7. Check that you are redirected to the Auth0 login page.
8. After you log in, check that you are redirected to the /dashboard route.
9. Log out via the top right kebab menu option.
10. Check that you are redirected to the /sign-in route and that the page loader message makes sense.
11. Click the "Sign In" button.
12. Check that you are redirected to the Auth0 login page.
13. After you log in, check that the page loader messages make sense.
14. Check that you are redirected to the /dashboard route.
15. Go to the "My Travel Requests" page via the top drop down nav and copy the URL.
16. Sign out from the kebab menu again.
17. This time paste the `http://localhost:8080/my-travel-requests` url into the URL bar to simulate direct navigation.
18. Check that you get a nice loader.
19. Check that you are redirected to the Auth0 login page.
20. After you log in, check that the page loader messages make sense.
21. Check that you are eventually redirected to http://localhost:8080/my-travel-requests
